kapitan/refs/secrets/vault.py

```diff
@@ -0,0 +1,167 @@
+    * VAULT_CAPATH: path to a directory of PEM-encoded CA cert files to verify the Vault server TLS certificate
+    * VAULT_NAMESPACE: specify the Vault Namespace, if you have one
+    """
+    env = {}
+    env['url'] = getenv( 'VAULT_ADDR', default='http://127.0.0.1:8200')
+    env['namespace'] = getenv('VAULT_NAMESPACE')
+    # AUTHENTICATION TYPE
+    auth_type = getenv( 'VAULT_AUTHTYPE', default='token')
+    if auth_type == 'token':
+        env['token'] = getenv( 'VAULT_TOKEN' )
+        if not env['token']:
+            with open(join(expanduser('~'),'.vault-token'),'r') as f:
+                env['token'] = f.read()
+    elif auth_type == 'userpass':
+        env['username'] = getenv( 'VAULT_USER' )
+        env['password'] = getenv( 'VAULT_PASSWORD' )
+    elif auth_type == 'approle':
+        env['role_id'] = getenv('VAULT_ROLE_ID')
+        env['secret_id'] = getenv( 'VAULT_SECRET_ID' )
+    # VERIFY VAULT SERVER TLS CERTIFICATE
+    verify = getenv( 'VAULT_SKIP_VERIFY', default='')
+    if verify.lower() == 'true':
+        env['verify'] = False
+    elif verify.lower() == 'false':
+        cert = getenv( 'VAULT_CACERT' )
+        if not cert:
+            cert_path = getenv( 'VAULT_CAPATH' )
+            if not cert_path:
+                raise Exception('Neither VAULT_CACERT nor VAULT_CAPATH specified')
+            env['verify'] = cert_path
+        else:
+            env['verify'] = cert
+    # CLIENT CERTIFICATE FOR TLS AUTHENTICATION
+    client_key,client_cert = getenv( 'VAULT_CLIENT_KEY' ), getenv( 'VAULT_CLIENT_CERT' ) 
+    if client_key != None and client_cert != None:
+        env['cert'] = (client_cert,client_key)
+    return env
+
+def vault_obj():
+    client = hvac.Client(
+        **{k:v for k,v in get_env().items() if v is not None}
+    )
+    assert (
+        client.is_authenticated()
+    ), "Vault Authentication Error, Environment Variables defined?"
+    return client
+
+
+class VaultSecret(Ref):
+
+    """
+    Hashicorp Vault can be used if using KV Secret Engine
+    """
+
+    def __init__(self,data,encode_base64=False,**kwargs):
+        """
+        set encoding_base64 to True to base64 encoding key before encrypting and writing
+        """
+        self._encrypt(data, encode_base64)
+        kwargs['encoding'] = self.encoding
+        super().__init__(self.data,**kwargs)
+        self.type_name = 'vault'
+
+    def _encrypt(self, data,encode_base64):
+        """
+        encrypt data
+        set encode_base64 to True to base64 encode data before writing
+        """
+        self.encoding = "original"
+        if encode_base64:
+            self.data = base64.b64encode(data.encode())
+            self.encoding = "base64"
+        else:
+            self.data = data.encode()
+
+
+    def reveal(self):
+        """
+        returns decrypted data
+        """
+        # can't use super().reveal() as we want bytes
+        ref_data = base64.b64decode(self.data)
```

Comment:
@daminisatya The refactoring in cli.py removing encode_base64 flag in older patch has addressed part of my concern. However, my concern is yet to be fixed to deal with various encoding schemes specifically and case when encoding scheme is wrongly set , otherwise can lead to potential bug scenario pointed out in 411 of cli.py (please see my latest comment) although setting up non-existing/unacceptable encoding scheme base65 still successfully decode using base64. When the reference controller holding the secret_obj during encode in cli.py is passed to the reveal in cli.py, the state seems to be available for the reveal(..) which has to be utilised for conditional based decoding based on encoding scheme. @adrianchifor  please clarify my understanding of the encrypt and reveal feature.